Content.Shared/RCD/Systems/RCDSystem.cs

```diff
@@ -35,313 +41,590 @@ public sealed class RCDSystem : EntitySystem
+            Loc.GetString("rcd-component-examine-mode-details", ("mode", Loc.GetString(component.CachedPrototype.SetName)));
+
+        args.PushMarkup(msg);
     }
 
-    private void OnAfterInteract(EntityUid uid, RCDComponent comp, AfterInteractEvent args)
+    private void OnAfterInteract(EntityUid uid, RCDComponent component, AfterInteractEvent args)
     {
         if (args.Handled || !args.CanReach)
             return;
 
         var user = args.User;
+        var location = args.ClickLocation;
 
-        TryComp<LimitedChargesComponent>(uid, out var charges);
-        if (_charges.IsEmpty(uid, charges))
+        // Initial validity checks
+        if (!_gameTiming.IsFirstTimePredicted)
+            return;
+
+        if (!location.IsValid(EntityManager))
+            return;
+
+        if (!TryGetMapGridData(location, out var mapGridData))
         {
-            _popup.PopupClient(Loc.GetString("rcd-component-no-ammo-message"), uid, user);
+            _popup.PopupClient(Loc.GetString("rcd-component-no-valid-grid"), uid, user);
             return;
         }
 
-        var location = args.ClickLocation;
-        // Initial validity check
-        if (!location.IsValid(EntityManager))
+        if (!IsRCDOperationStillValid(uid, component, mapGridData.Value, args.Target, args.User))
             return;
 
-        var gridId = location.GetGridUid(EntityManager);
-        if (!HasComp<MapGridComponent>(gridId))
+        if (!_net.IsServer)
+            return;
+
+        // Get the starting cost, delay, and effect from the prototype
+        var cost = component.CachedPrototype.Cost;
+        var delay = component.CachedPrototype.Delay;
+        var effectPrototype = component.CachedPrototype.Effect;
+
+        #region: Operation modifiers
+
+        // Deconstruction modifiers
+        if (component.CachedPrototype.Mode == RcdMode.Deconstruct)
         {
-            location = location.AlignWithClosestGridTile();
-            gridId = location.GetGridUid(EntityManager);
-            // Check if fixing it failed / get final grid ID
-            if (!HasComp<MapGridComponent>(gridId))
-                return;
+            // Deconstructing an object
+            if (args.Target != null)
+            {
+                if (TryComp<RCDDeconstructableComponent>(args.Target, out var destructible))
+                {
+                    cost = destructible.Cost;
+                    delay = destructible.Delay;
+                    effectPrototype = destructible.Effect;
+                }
+            }
+
+            // Deconstructing a tile
+            else
+            {
+                var tile = _mapSystem.GetTileRef(mapGridData.Value.GridUid, mapGridData.Value.Component, mapGridData.Value.Location);
+                var protoName = (tile.GetContentTileDefinition().ID != "Lattice") ? "DeconstructTile" : "DeconstructLattice";
```

Comment:
you really ought to just put a field on content tile definition because this is horrendously hardcoded.